massa-network/src/peer_info_database.rs

```diff
@@ -823,19 +738,25 @@ impl PeerInfoDatabase {
                 ));
             }
         }
+        let peer = if let Some(p) = self.get_peer_type(ip) {
+            p
+        } else {
+            let p = PeerInfo::new(*ip, false);
+            self.peers.insert(*ip, p);
+            self.get_peer_type(ip).unwrap()
+        };
+
+        if self.is_max_in_connection_count_reached(peer) {
+            return Err(NetworkError::PeerConnectionError(
+                NetworkConnectionErrorType::MaxPeersConnectionReached(*ip),
+            ));
+        }
 
         let res = match self.peers.entry(*ip) {
             hash_map::Entry::Occupied(mut occ) => {
```

Comment:
a few lines above, when the peer is not found you do `self.peers.insert(*ip, p);`  so this will always be occupied and never Vacant

Reply:
I will re-review this whole function when this part is corrected